README.txt
ISEE-3/ICE Telemetry demodulator and decoder
Copyright Phil Karn, KA9Q, 28 June 2014
version 0.10
May be used under the terms of the GNU Public License 2.0 (GPL)

The major change in 0.10 is that "decode" now uses hybrid Fano/Viterbi decoding. By default it first tries Fano; if this fails *and* the previous frame successfully decoded (with either algorithm), decoding is tried again with the Viterbi algorithm.

This means that after a sustained period of noise (or at the beginning of a recording) at least one frame must be successfully decoded with Fano before Viterbi will be used at all. This is to keep the program from falling way behind when fed noise. (You can force Viterbi decoding on all frames with the -V option. Likewise you can disable it and do only Fano decoding with the -F option.)


The major change in 0.9 was the new program "decode" that replaces both vdecode and framer. It also incorporates the Fano sequential decoding algorithm as an alternative to the slow Viterbi algorithm. Fano is enormously faster than Viterbi but performs about 0.5 - 1.0 dB worse. Fano also has a variable execution speed that increases sharply as the decoder Eb/No threshold of about 2.5 dB is reached.

This is a snapshot of my ISEE-3/ICE telemetry decoder. It is now broken into three (not four) modules that you run in a UNIX pipeline like this:

pmdemod input_file | symdemod [-c <symrate>] | decode [-f] [-r <symrate>]

The -f option to decode specifes the Fano algorithm; the default is Viterbi.

Note that with ISEE-3, the symrate is always twice the data rate.

Starting with version 0.8, pmdemod can read from a pipe or device as well as a file, and will read from standard input if no file name is given as a parameter.

The input stream is expected to be a series of 16-bit signed integer samples in little-endian format. The I channel is expected first. If Q is first, use the -f (flip) flag to pmdemod to invert the spectrum.

Each program writes its status to stderr so you can see what it's doing while its output is redirected. Use the -q option to a program to shut up its diagnostics.

This version has been tested at 32 sps/16 bps, 128 sps/64 bps, 1024 sps/512 bps and 4096 sps/2048 bps. Specifying an integer to the -c option of symdemod causes the actual clock frequency (which is slightly higher) to be used. The measured clock speeds (at -3 km/s velocity) are approximately:

1024 sps: 1024.475 Hz
4096 sps: 4097.9 Hz
32 sps: 32.0148 Hz

pmdemod - reads raw receiver file, tracks carrier, writes baseband PM samples (16-bit signed ints) on stdout

-S starting carrier frequency estimate, Hz
-W +/- Frequency search range while locked, Hz (searches entire passband when unlocked)
-D Doppler chirp rate, Hz/s
-t C/N0 lock threshold, dB; default 21 dB (good for 512 bps; should be lowered for lower data rates)
-q Quiet mode
-b carrier tracking FFT bin size, Hz; default 4 Hz (changed from 1 Hz to better track spin Doppler)
-f flip I & Q samples (invert spectrum)
-r Sample rate, Hz; default: 250000

symdemod - reads output of pmdemod on stdin, writes demodulated, 8-bit (offset-128) soft decisions on stdout
-c Symbol rate, Hz; default 1024.475 (measured rate at 1024 Hz nominal; divide or multiply by 2^n)
-C Clocks per symbol (default 1 for symbol rates > 1000; 1024/symrate for lower rates)
-r Sample rate, Hz; default 250000
-q quiet mode

decode - Reads output of symdemod on stdin, writes decoded frames on stdout
-V Use Viterbi decoding only (slow - only runs about 230 bps on a typical i7)
-F Use Fano sequential decoding only
-r Symbol rate, Hz; default 1024 (used only for output timestamps)
-s Fano scale parameter; default 8
-d Fano delta parameter; default 32
-m Fano maximum cycles per bit; default 100

The new 'decode' program replaces both vdecode and framer, which were used as follows:

pmdemod input_file | symdemod -c <symrate> | vdecode | framer -r <bitrate>

vdecode - Reads output of symdemod on stdin, writes Viterbi-decoded bits on stdout ('0'/'1')
-p Start with opposite decoder symbol phase (also flips automatically based on encoded sync observations)
-i Status update interval, bits; default 1024
-d Viterbi decode delay, bits; default 200
-q quiet mode

framer - reads output of vdecode on stdin, detects frame sync, writes decoded telemetry frames in hex on stdout
-r bitrate, Hz; default 512 (only for time estimation)

The old programs icedemod, icesync and bitsync are included for reference but are not built by default and may not even still compile.



CHANGES
Change log for the KA9Q ISEE-3/ICE demodulator/decoder software
Phil Karn, karn@ka9q.net
Last updated 10 July 2014

Version 0.11:

This is mainly system performance tuning of the 'decode' program. Unless it is always selected (-V), memory for the Viterbi decoder is now allocated only when actually needed for fallback from a failed Fano decode attempt. The Viterbi decoder uses about 1 GB, so this might make a difference on smaller machines.

The -p ("persistent") option has been added to 'decode' for hybrid mode to force fallback to Viterbi whenever Fano fails to decode a frame, instead of only when the previous frame succeeded. Be careful: this will cause the decoder to fall behind real time on no signal when the system is not fast enough for real-time Viterbi decoding (which is usually the case at 512 bps and up.)

Versions of decode (decode_port) and the obsolete vdecode (vdecode_port) that use a Viterbi decoder in portable C are included for non-x86/SSE2 systems. Warning: this is very slow -- about 30 bps on an i7!

Version 0.10:

'decode' uses hybrid decoding by default. A Fano (sequential) decoder is tried first. If it fails to decode, a Viterbi decode is tried IF the previous frame decoded successfully (with either Viterbi or Fano). This is a heuristic intended to recover the occasional weak frame without trying to Viterbi-decode pure noise, which would cause most systems to fall behind real-time.

This means that after a sustained period of noise (or at the beginning of a recording) at least one frame must be successfully decoded with Fano before Viterbi will be used at all. This is to keep the program from falling way behind when fed noise. (You can force Viterbi decoding on all frames with the -V option. Likewise you can disable it and do only Fano decoding with the -F option. Also see the -p option added to version 0.11 above.)

Version 0.9:

The new program "decode" replaces both vdecode and framer. It also incorporates the Fano sequential decoding algorithm as an alternative to the slow Viterbi algorithm. Fano is enormously faster than Viterbi but performs about 0.5 - 1.0 dB worse. Fano also has a variable execution speed that increases sharply as the decoder Eb/No threshold of about 2.5 dB is reached.

Version 0.8:

'pmdemod' can read from a pipe (e.g., on standard input); previous versions could only read from a file.

Versions 0.6-0.7:

These incorporate various performance enhancements, particularly in the symbol demodulator (symdemod), which is now vastly faster at the lower symbol rates due to an improved symbol timing algorithm. Because it uses a fairly stiff tracking loop that may not acquire if the initial estimate is too far off, symdemod now knows the actual clock frequency of the spacecraft (which is a little high) and uses it unless the frequency you give the -c option has a decimal point in it.

Version 0.5:

This is  a minor change from 0.4. It includes a change made for Tim R to get the Viterbi decoder to compile on a pre-SSE-4.1 system, plus some updates to the README.txt giving the measured symbol rates for the symdemod.

The -C option to symdemod specifies the number of clocks per symbol. This is set automatically to 1 for symbol rates above 1000 Hz (thus allowing for a -c parameter < 1024 should it drop that low after perigee due to Doppler) and to 1024/symrate for 512 sps and lower. I.e., at 32 sps (16 data bits/sec) symdemod will automatically assume 1024/32 = 32 cycles of 1024 Hz subcarrier per symbol.

This code has been operationally tested at all three data rates I've seen so far:

1024 sps/512 bps (2 sec frames)
4096 sps/2048 bps (1/2 sec frames)
32 sps/16 bps    (64 sec frames)

baudline's observation of the spin-induced Doppler on the RF carrier (caused by a 4.75 cm offset of the S-band antenna from the spacecraft c.g.) made me re-evaluate my carrier frequency tracking algorithm in pmdemod. I had been using a rather narrow FFT frequency bin (analogous to a carrier PLL loop bandwidth) of only 1 Hz, and this caused an appreciable amount of phase twist as the spacecraft rotated 2 radians in each integration period.

I've increased the default to 5 Hz, which should still provide adequate SNR for carrier tracking especially when the new optional search range feature is used. Here's how it works: when the C/No is above a specified threshold (default 21 dB-Hz, appropriate for 512 bps but should be lowered for lower data rates), the carrier is declared "locked" and only a specified range of frequencies are searched around it to minimize the risk of being distracted by a temporary noise spike somewhere else in the receiver passband. When the C/No falls below threshold, the algorithm reverts to looking for the strongest bin anywhere in the passband. This makes essentially no difference when C/No is high (since there's little risk of any noise exceeding the true carrier) but it might make a difference when operating at low C/No, such as at low data rates on smaller antennas.

--Phil


bitsync.c
// Demodulate and decode ISEE-3 frames
// Phil Karn, KA9Q, June 2014

#include <stdio.h>
#include <linux/limits.h>
#include <unistd.h>
#include <stdlib.h>
#define __USE_GNU   1
#include <math.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <assert.h>
#include <locale.h>
#include "code.h"
#include "viterbi224.h"
#include "timeformat.h"

#define FRAMEBITS 1024   // 1024 bits per minor frame
#define FRAMESYMBOLS (2*FRAMEBITS)
#define SYNCBITS 34      // Last 34 bits of encoded tail + sync are usable
#define SYMRATE 1024.467    // Nominal symrate for 512 bps mode

#define SYNCWORD 0x12fc819fbeLL

int Verbose;
double Symbolsamples;
double Samprate = 250000;   // Sample rate of PM demodulated symbols
double Symrate = SYMRATE;   // Symbol rate

double trial_demod(short *samples,int firstsample,double symbolsamples,int symbols);

// 34 bit encoded sync symbols

int sync_vector[] = {
  0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1,
  1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};


int main(int argc,char *argv[]){
  int i,fd;
  char *filename;
  struct stat statbuf;
  off_t length;
  short *samples;
  long long nsamples;
  int firstsample;
  char *locale;
  double maxenergy;
  int symphase;
  int nochange;
  double clock_incr;
  int phase_incr;
  double window = 1; // Seconds of symbols to examine
  int ind;
  double scount;
  void *vd;
  unsigned char vdsyms[2];
  int renorms = 0;
  int symbols = 0;
  double symbol_register[FRAMESYMBOLS];
  double sync_peak_even,sync_peak_odd,sync_sum;
  int k;
  int vd_phase;
  double integrator;
  int frame = 1;
  unsigned long long decoded_data[16];
  double gain;
  size_t offset = 0;

  memset(symbol_register,0,sizeof(symbol_register));
  memset(decoded_data,0,sizeof(decoded_data));
  if((locale = getenv("LANG")) != NULL)
    setlocale(LC_ALL,locale);
  else
    setlocale(LC_ALL,"en_US.utf8");

  offset = 0;
  Samprate = 250000;
  Symrate = SYMRATE;
  while((i = getopt(argc,argv,"c:r:o:s:")) != EOF){
    switch(i){
    case 'o':
      offset = atoi(optarg);    // Skip to sample
      break;
    case 'c':
      Symrate = atof(optarg);   // Initial estimate, will be updated
      break;
    case 'r':
      Samprate = atof(optarg);  // Fixed, assumed exact
      break;
    case 's':
      Symrate = atof(optarg);   // Symbol rate
      break;
    }
  }
Symbolsamples = Samprate/Symrate;

  // Read baseband samples, downsampled to Samprate (important!)                                                                                      
  filename = argv[optind];
  if(lstat(filename,&statbuf) == -1){
    fprintf(stderr,"lstat(%s) failed: %s\n",filename,strerror(errno));
    exit(1);
  }
  if(!S_ISREG(statbuf.st_mode)){
    fprintf(stderr,"%s is not an ordinary file\n",filename);
    exit(1);
  }
  length = statbuf.st_size;
 if((fd = open(filename,O_RDONLY)) == -1){
    fprintf(stderr,"open(%s,readonly) failed; %s\n",filename,strerror(errno));
    exit(1);
  }
  if((samples = mmap(NULL,length,PROT_READ,MAP_SHARED,fd,0)) == MAP_FAILED){
    fprintf(stderr,"mmap(%s,%lld) failed: %s\n",filename,
            (long long)length,strerror(errno));
    close(fd);
    exit(1);
  }
  nsamples = length / sizeof(*samples);
  printf("%s: %'lld samples; %'.3lf sec (%s) @ %'.1lf Hz\n",
	 filename,nsamples,nsamples/Samprate,format_hms(nsamples/Samprate),Samprate);
  vd = create_viterbi224(250); // Must be bigger than the traceback depth
  assert(vd != NULL);
  init_viterbi224(vd,0); // See what happens

  // Look at blocks of data 'window' seconds wide
  // A 1 second window gives a SNR of 10*log10(Symrate) = ~30 dB over the symbol SNR
  for(firstsample = offset + Symbolsamples/2;
      firstsample + FRAMESYMBOLS*Symbolsamples < nsamples;
      firstsample += FRAMESYMBOLS*Symbolsamples){

    // Full search of symbol phase at current clock estimate
    maxenergy = 0;
    symphase = 0; // Avoid uninitialized variable warning
    for(i = -Symbolsamples/2; i < Symbolsamples/2; i++){
      double energy;
      
      energy = trial_demod(samples,firstsample+i,Symbolsamples,FRAMESYMBOLS); // demod over one window
      if(energy > maxenergy){
	maxenergy = energy;
	symphase = i;
      }
    }
    firstsample += symphase;
    symphase = 0;
    // Fine-tune the symbol clock and phase estimates
    clock_incr = 0.5 * Symbolsamples/(window*Samprate); // +/-0.5 sample at end of window
    phase_incr = 1;                       // +/-1 sample throughout window
    for(nochange = 0; nochange < 2;){
      double energy;
      
      // Adjust clock, see if it improves
      if((energy = trial_demod(samples,firstsample,Symbolsamples + clock_incr,FRAMESYMBOLS)) > maxenergy){
	maxenergy = energy;
	Symbolsamples += clock_incr;
	Symrate = Samprate / Symbolsamples;
	nochange = 0;        // we made progress, don't stop
      } else if((energy = trial_demod(samples,firstsample,Symbolsamples - clock_incr,FRAMESYMBOLS)) > maxenergy){
	maxenergy = energy;
	Symbolsamples -= clock_incr;
	Symrate = Samprate/Symbolsamples;
	clock_incr = -clock_incr; // Try this same direction first next time
	nochange = 0;
      } else
	nochange++;          // no improvement
      
      // See if changing phase helps
      if((energy = trial_demod(samples,firstsample + phase_incr,Symbolsamples,FRAMESYMBOLS)) > maxenergy){
	maxenergy = energy;
	firstsample += phase_incr;
	nochange = 0;
      } else if((energy = trial_demod(samples,firstsample - phase_incr,Symbolsamples,FRAMESYMBOLS)) > maxenergy){
	maxenergy = energy;
	firstsample += phase_incr;
	phase_incr = -phase_incr; // Try this direction first next time
	nochange = 0;
      } else
	nochange++;           // If neither a change in clock or phase helped, we'll fall through the loop
    }
    printf("Frame %'d starting at sample %'d (%'.3lf sec, %s): clock %'.4lf Hz; %'.4lf samp/sym; energy %.3lf dB\n",
	    frame,firstsample,firstsample/Samprate,format_hms(firstsample/Samprate),Symrate,Symbolsamples,10*log10(maxenergy));

    fflush(stdout);
    // Demodulate using parameters
    ind = firstsample;       // Index of first sample to be integrated
    scount = ind + 0.5 * Symbolsamples; // Sample at middle of first symbol; note integer truncation
    sync_peak_even = sync_peak_odd = 0;
    for(i=0;i<FRAMESYMBOLS;i++){
      // Integrate first half of symbol
      integrator = 0;        // reset integrator
      for(;ind < scount;ind++)
	integrator -= samples[ind];

      // Integrate second half of symbol
      scount += 0.5 * Symbolsamples;   
      for(; ind < scount;ind++)
	integrator += samples[ind];
      scount += 0.5 * Symbolsamples; // update for next symbol

      symbols++;
      // Accumulate in shift register
      symbol_register[i] = integrator;
      // Run sync correlator
      sync_sum = 0;
      for(k=0;k<34;k++){
	if(sync_vector[k])
	  sync_sum += symbol_register[(FRAMESYMBOLS + i - 34 + k)% FRAMESYMBOLS];
	else
	  sync_sum -= symbol_register[(FRAMESYMBOLS + i - 34 + k)% FRAMESYMBOLS];
      }	  
      if((i % 2) == 0){
	if(sync_sum > sync_peak_even)
	  sync_peak_even = sync_sum;
      } else {
	if(sync_sum > sync_peak_odd)
	  sync_peak_odd = sync_sum;
      }
    }
    // Use the best sync peak just to determine Viterbi decoder phase
    vd_phase = sync_peak_even < sync_peak_odd;
    //    printf("vd phase = %d\n",vd_phase);
    
    gain = 75./sqrt(maxenergy); // Hack
    for(i=0;i<FRAMESYMBOLS;i++){
      // Integrator now has soft decision sym, apply gain and bias for viterbi decoder and clip
      integrator = gain * symbol_register[i] + 128;
      if(integrator > 255)
	integrator = 255;
      else if(integrator < 0)
	integrator = 0;

      vdsyms[vd_phase] = integrator;
      if(vd_phase == 1){
	int bit;
	renorms += update_viterbi224_blk(vd,vdsyms,1);

	// Chain back and decode a bit
	// This is far enough back for the paths to merge, we don't
	// need to waste time finding the best path
	bit = decodebit_viterbi224(vd,200,0); // Cannot be larger than value given to create_viterbi
	// 1024-bit shift register with decoded bits
	for(k=0;;k++){
	  decoded_data[k] <<= 1;
	  if(k == 15)
	    break;
	  decoded_data[k] |= (decoded_data[k+1] >> 63);
	}
	decoded_data[15] |= bit;
	
	//	printf("symbols %6u renorms %02d data %016llx\n",i,renorms,decoded_data[15]);
	if((decoded_data[15] & 0xffffffffffL) == SYNCWORD){
	  int k;
	  
	  for(k=0;k<16;k++){
	    int n;
	    
	    for(n=56;n >= 0;n -= 8){
	      printf("%02llx",(decoded_data[k] >> n) & 0xff);
	      if(n == 0 && (k % 2) == 1)
		putchar('\n');
	      else
		putchar(' ');
	    }
	  }
	}
      }
      vd_phase = !vd_phase;
    }
    putchar('\n');
    //    printf("sync peak even %lf odd %lf\n",sync_peak_even,sync_peak_odd);
    frame++;
    fflush(stdout);
  }
  exit(0);
}

double trial_demod(short *samples,int firstsample,double symbolsamples,int symbols){
  double energy,integrator;
  int ind,i;
  double scount;

  energy = 0;
  ind = firstsample;       // Index of first sample to be integrated
  scount = ind + 0.5 * symbolsamples; // Sample at middle of first symbol; note integer truncation

  for(i=0;i<symbols;i++){
    integrator = 0;        // reset integrator
    // Integrate first half of symbol
    for(;ind < scount;ind++)
      integrator -= samples[ind];

    // Integrate second half of symbol
    scount += 0.5 * symbolsamples;   
    for(; ind < scount;ind++)
      integrator += samples[ind];

    scount += 0.5 * symbolsamples; // update for next symbol

    integrator *= integrator; // symbol amplitude -> energy
    energy += integrator;     // Accumulate energy
  }
  //  return energy / (ind - firstsample); // Normalize for number of samples used
  return energy / symbols;
}


